ts/index.ts

```diff
@@ -280,49 +281,92 @@ class Gossipsub extends BasicPubsub {
    * @param {Peer} peer
    * @param {Message} msg
    */
-  _processRpcMessage (peer: Peer, msg: InMessage): void {
+  async _processRpcMessage (peer: Peer, msg: InMessage): Promise<void> {
     const msgID = this.getMsgId(msg)
 
     // Ignore if we've already seen the message
     if (this.seenCache.has(msgID)) {
+      this.score.duplicateMessage(peer.id.toB58String(), msg as Message)
       return
     }
     this.seenCache.put(msgID)
 
-    super._processRpcMessage(peer, msg)
+    this.score.validateMessage(peer.id.toB58String(), msg as Message)
+    await super._processRpcMessage(peer, msg)
+  }
+
+  /**
+   * Publish a message sent from a peer
+   * @override
+   * @param {Peer} peer
+   * @param {InMessage} msg
+   * @returns {void}
+   */
+  _publishFrom (peer: Peer, msg: InMessage): void {
+    this.score.deliverMessage(peer.id.toB58String(), msg as Message)
     const topics = msg.topicIDs
 
     // If options.gossipIncoming is false, do NOT emit incoming messages to peers
-    if (!this._options.gossipIncoming) {
-      return
-    }
-
-    // Emit to floodsub peers
-    this.peers.forEach((peer) => {
-      if (peer.protocols.includes(constants.FloodsubID) &&
-        peer.id.toB58String() !== msg.from &&
-        utils.anyMatch(peer.topics, topics) &&
-        peer.isWritable
-      ) {
-        peer.sendMessages(utils.normalizeOutRpcMessages([msg]))
-        this.log('publish msg on topics - floodsub', topics, peer.id.toB58String())
-      }
-    })
+    if (this._options.gossipIncoming) {
```

Comment:
I would rather short circuit here than have all the remaining lines with more indentation. What do you think?

Reply:
The reason I changed this is to still allow for `super._publishFrom` to be called at the bottom of the function, regardless of `gossipIncoming`, L340 below.
I forgot to put this in the description above, but its a slight change in behavior.
The reason for calling `super._publishFrom` is to have this ordering, which was switched before:
1. (optionally) publish incoming message to floodsub/gossipsub peers (`this._publishFrom`)
2. (optionally) emit incoming message as an event (`super._publishFrom`)

Reasoning for this is that iirc, event handlers are executed synchronously as events are emitted, which seems like it might interfere with the timing of publishing incoming messages. I guess the general rule I've been using is to try to have events emitted at the end of processing.

Reply:
it makes sense!